docs/specs/m2m.md

```diff
@@ -0,0 +1,1152 @@
+}
+```
+
+The Token Exchange request:
+
+```
+POST /oauth2/token HTTP/1.1
+Content-Type: application/x-www-form-urlencoded
+
+grant_type=urn:ietf:params:oauth:grant-type:token-exchange
+resource=https://onlinestore.myapp.com
+requested_token_type=urn:ietf:params:oauth:token-type:access_token
+subject_token=SUBJECT_TOKEN
+subject_token_type=urn:ietf:params:oauth:token-type:access_token
+actor_token=ACTOR_TOKEN
+actor_token_type=urn:ietf:params:oauth:token-type:access_token
+```
+
+The exchanged token:
+
+```json
+{
+  "iss": "https://auth.myapp.com",
+  "sub": "johndoe",
+  "aud": ["https://onlinestore.myapp.com"],
+  "client_id": "inventory",
+  "https://authgear.com/claims/user/roles": ["onlinestore:admin"],
+  "scope": "read:orders write:orders delete:orders"
+}
+```
+
+Note that
+
+- The `scope` of `subject_token` is `openid offline_access`.
+- The `aud` of `subject_token` is `https://auth.myapp.com`, which is not acceptable by `https://onlinestore.myapp.com`.
+- The `scope` of `actor_token` is `read:orders`, so `https://inventory.myapp.com` does not have admin access on its own.
+- The `scope` of exchanged token is `read:orders write:orders delete:orders`. It means that when `https://inventory.myapp.com` acting on behalf of `johndoe`, who is `onlinestore:admin`, has additional `scope` inherit from `sub` (`johndoe`).
+
+### Discussion: (Auth0) Rich Authorization Requests
+
+[RFC9396](https://datatracker.ietf.org/doc/html/rfc9396) introduces `authorization_details`,
+which is an enhancement over `scope` and `resource` to specify authorization details in a structured way in form of JSON.
+
+This is supported by Auth0 as an Add-on. See https://auth0.com/docs/get-started/apis/configure-rich-authorization-requests
+
+### Discussion: (Future work) Restricting access to Admin API
+
+In Auth0, `RFC6749 section-4.4` is also used to create an `access_token` that can be used to access the Management API.
+
+We can implement the same for our Admin API.
+Specifically, we need to introduce an artificial Resource with URI `https://auth.myapp.com/_api/admin`, and define a comprehensive list of scopes that restrict access to the different Resources within the Admin API.
+
+## MVP
+
+In the MVP, we are going to implement M2M only.
+This means
+
+- The table `__auth_oauth_authorization_resource` **IS NOT** added.
+- The `resource` parameter in the Authorization endpoint **IS NOT** implemented.
+- The grant_type `client_credentials` **IS** supported in the Token endpoint.
+- The `resource` parameter and `scope` parameter **ARE** supported when `grant_type=client_credentials`.
+- The relevant CRUD of Resource, Scope, and Client **ARE** supported.
+
+### Changes in data models
+
+Here are the schema of the changes:
+
+```sql
+CREATE TABLE _auth_resource (
+  id text PRIMARY KEY,
+  app_id text NOT NULL,
+  created_at timestamp without time zone NOT NULL,
+  updated_at timestamp without time zone NOT NULL,
+  uri text NOT NULL,
+  name text,
+  data jsonb
```

Comment:
@louischan-oursky What is this data for?

Reply:
Maybe I should call this `metadata`. It is a column to store metadata about the entity, like the recent `_auth_user.metadata`